File: samples/Cpp/TestCpp/Classes/ExtensionsTest/CocoStudioSceneTest/TriggerCode/cons.cpp

#include "cons.h"

using namespace cocos2d;
using namespace cocos2d::extension;


IMPLEMENT_CLASS_INFO(TimeElapsed)

TimeElapsed::TimeElapsed(void)
:_fTotalTime(0.0f)
,_fTmpTime(0.0f)
,_pScheduler(NULL)
,_bSuc(false)
{
	_pScheduler = CCDirector::sharedDirector()->getScheduler();
	_pScheduler->retain();
}

TimeElapsed::~TimeElapsed(void)
{
	CC_SAFE_RELEASE(_pScheduler);
}

bool TimeElapsed::init()
{
	_pScheduler->scheduleSelector(schedule_selector(TimeElapsed::update), this, 0.0f , kCCRepeatForever, 0.0f, false);
    return true;
}

bool TimeElapsed::detect()
{
    return _bSuc;
}

void TimeElapsed::serialize(const rapidjson::Value &val)
{
    int count = DICTOOL->getArrayCount_json(val, "dataitems");
	for (int i = 0; i < count; ++i)
	{
		const rapidjson::Value &subDict = DICTOOL->getSubDictionary_json(val, "dataitems", i);
		std::string key = DICTOOL->getStringValue_json(subDict, "key");
		if (key == "TotalTime")
		{
			_fTotalTime = DICTOOL->getFloatValue_json(subDict, "value");
		}
	}
}

void TimeElapsed::serialize(cocos2d::extension::CocoLoader *pCocoLoader, cocos2d::extension::stExpCocoNode *pCocoNode)
{
	int length = pCocoNode->GetChildNum();
	int count = 0;
	stExpCocoNode *pDataItemsArray = pCocoNode->GetChildArray(pCocoLoader);
	std::string key;
	const char *str = NULL;
	for (int i = 0; i < length; ++i)
	{
		count = pDataItemsArray[i].GetChildNum();
		stExpCocoNode *pDataItemArray = pDataItemsArray[i].GetChildArray(pCocoLoader);
		key = pDataItemArray[0].GetValue(pCocoLoader);
		str = pDataItemArray[1].GetValue(pCocoLoader);
		if (key == "TotalTime")
		{
			if (str != NULL)
			{
				_fTotalTime = atof(str); //DICTOOL->getFloatValue_json(subDict, "value");
			}
		}
	}
}

void TimeElapsed::removeAll()
{
	_pScheduler->unscheduleUpdateForTarget(this);
}

void TimeElapsed::update(float dt)
{
	_fTmpTime += dt;
	if (_fTmpTime > _fTotalTime)
	{
		_fTmpTime = 0.0f;
		_bSuc = true;
	}
}

IMPLEMENT_CLASS_INFO(ArmatureActionState)
ArmatureActionState::ArmatureActionState(void)
: _nTag(-1)
, _nState(-1)
, _bSuc(false)
{
}

ArmatureActionState::~ArmatureActionState(void)
{
}

bool ArmatureActionState::init()
{
	do 
	{
		CCNode *pNode = SceneReader::sharedSceneReader()->getNodeByTag(_nTag);
		CC_BREAK_IF(pNode == NULL);
		CCComRender *pRender = (CCComRender*)(pNode->getComponent(_comName.c_str()));
		CC_BREAK_IF(pRender == NULL);
		CCArmature *pAr = (CCArmature *)(pRender->getNode());
		CC_BREAK_IF(pAr == NULL);
		TriggerMng::getInstance()->addArmatureMovementCallBack(pAr, this, movementEvent_selector(ArmatureActionState::animationEvent));
	} while (0);
	
	return true;
}

bool ArmatureActionState::detect()
{
    return _bSuc;
}

void ArmatureActionState::serialize(const rapidjson::Value &val)
{
	int count = DICTOOL->getArrayCount_json(val, "dataitems");
	for (int i = 0; i < count; ++i)
	{
		const rapidjson::Value &subDict = DICTOOL->getSubDictionary_json(val, "dataitems", i);
		std::string key = DICTOOL->getStringValue_json(subDict, "key");
		if (key == "Tag")
		{
			_nTag = DICTOOL->getIntValue_json(subDict, "value");
			continue;
		}
		else if (key == "componentName")
		{
			_comName = DICTOOL->getStringValue_json(subDict, "value");
			continue;
		}
		else if (key == "AnimationName")
		{
			_aniname = DICTOOL->getStringValue_json(subDict, "value");
			continue;
		}
		else if (key == "ActionType")
		{
			_nState = DICTOOL->getIntValue_json(subDict, "value");
			continue;
		}
	}
}

void ArmatureActionState::serialize(cocos2d::extension::CocoLoader *pCocoLoader, cocos2d::extension::stExpCocoNode *pCocoNode)
{
	int length = pCocoNode->GetChildNum();
	int count = 0;
	stExpCocoNode *pDataItemsArray = pCocoNode->GetChildArray(pCocoLoader);
	std::string key;
	const char *str = NULL;
	for (int i = 0; i < length; ++i)
	{
		count = pDataItemsArray[i].GetChildNum();
		stExpCocoNode *pDataItemArray = pDataItemsArray[i].GetChildArray(pCocoLoader);
		key = pDataItemArray[0].GetValue(pCocoLoader);
		str = pDataItemArray[1].GetValue(pCocoLoader);
		if (key == "Tag")
		{
			if (str != NULL)
			{
				_nTag = atoi(str);//DICTOOL->getIntValue_json(subDict, "value");
			}
		}
		else if (key == "componentName")
		{
			if (str != NULL)
			{
				_comName = str; //DICTOOL->getStringValue_json(subDict, "value");
			}
			
		}
		else if (key == "AnimationName")
		{
			if (str != NULL)
			{
				_aniname = str; //DICTOOL->getStringValue_json(subDict, "value");
			}
		}
		else if (key == "ActionType")
		{
			if (str != NULL)
			{
				_nState = atoi(str); //DICTOOL->getIntValue_json(subDict, "value");
			}
		}
	}
}

void ArmatureActionState::removeAll()
{
	do 
	{
		CCNode *pNode = SceneReader::sharedSceneReader()->getNodeByTag(_nTag);
		CC_BREAK_IF(pNode == NULL);
		CCComRender *pRender = (CCComRender*)(pNode->getComponent(_comName.c_str()));
		CC_BREAK_IF(pRender == NULL);
		CCArmature *pAr = (CCArmature *)(pRender->getNode());
		CC_BREAK_IF(pAr == NULL);
		TriggerMng::getInstance()->removeArmatureMovementCallBack(pAr, this, movementEvent_selector(ArmatureActionState::animationEvent));
	} while (0);
}

void ArmatureActionState::animationEvent(cocos2d::extension::CCArmature *armature, cocos2d::extension::MovementEventType movementType, const char *movementID)
{
	 std::string id = movementID;
	if (movementType == _nState && id.compare(_aniname) == 0)
	{
		_bSuc = true;
	}
}

IMPLEMENT_CLASS_INFO(NodeInRect)
NodeInRect::NodeInRect(void)
:_nTag(-1)
{
}

NodeInRect::~NodeInRect(void)
{
}

bool NodeInRect::init()
{

    return true;
}

bool NodeInRect::detect()
{
	CCNode *pNode = SceneReader::sharedSceneReader()->getNodeByTag(_nTag);
	if (pNode != NULL && fabs(pNode->getPositionX() - _origin.x) <= _size.width && fabs(pNode->getPositionY() - _origin.y) <= _size.height)
	{
		return true;
	}
    return false;
}

void NodeInRect::serialize(const rapidjson::Value &val)
{
	int count = DICTOOL->getArrayCount_json(val, "dataitems");
	for (int i = 0; i < count; ++i)
	{
		const rapidjson::Value &subDict = DICTOOL->getSubDictionary_json(val, "dataitems", i);
		std::string key = DICTOOL->getStringValue_json(subDict, "key");
		if (key == "Tag")
		{
			_nTag = DICTOOL->getIntValue_json(subDict, "value");
			continue;
		}
		else if (key == "originX")
		{
			_origin.x = DICTOOL->getIntValue_json(subDict, "value");
			continue;
		}
		else if (key == "originY")
		{
			_origin.y = DICTOOL->getIntValue_json(subDict, "value");
			continue;
		}
		else if (key == "sizeWidth")
		{
			_size.width = DICTOOL->getIntValue_json(subDict, "value");
			continue;
		}
		else if (key == "sizeHeight")
		{
			_size.height = DICTOOL->getIntValue_json(subDict, "value");
			continue;
		}
	}
}

void NodeInRect::serialize(cocos2d::extension::CocoLoader *pCocoLoader, cocos2d::extension::stExpCocoNode *pCocoNode)
{
	int length = pCocoNode->GetChildNum();
	int count = 0;
	stExpCocoNode *pDataItemsArray = pCocoNode->GetChildArray(pCocoLoader);
	std::string key;
	const char *str = NULL;
	for (int i = 0; i < length; ++i)
	{
		count = pDataItemsArray[i].GetChildNum();
		stExpCocoNode *pDataItemArray = pDataItemsArray[i].GetChildArray(pCocoLoader);
		key = pDataItemArray[0].GetValue(pCocoLoader);
		str = pDataItemArray[1].GetValue(pCocoLoader);
		if (key == "Tag")
		{
			if (str != NULL)
			{
				_nTag = atoi(str);//DICTOOL->getIntValue_json(subDict, "value");
			}
		}
		else if (key == "originX")
		{
			if (str != NULL)
			{
				_origin.x = atoi(str); //DICTOOL->getIntValue_json(subDict, "value");
			}
		}
		else if (key == "originY")
		{
			if (str != NULL)
			{
				_origin.y = atoi(str); //DICTOOL->getIntValue_json(subDict, "value");
			}
		}
		else if (key == "sizeWidth")
		{
			if (str != NULL)
			{
				_size.width = atoi(str); //DICTOOL->getIntValue_json(subDict, "value");
			}
		}
		else if (key == "sizeHeight")
		{
			if (str != NULL)
			{
				_size.height = atoi(str); //DICTOOL->getIntValue_json(subDict, "value");
			}
		}
	}
}

void NodeInRect::removeAll()
{
	CCLOG("NodeInRect::removeAll");
}


IMPLEMENT_CLASS_INFO(NodeVisible)
NodeVisible::NodeVisible(void)
: _nTag(-1)
, _bVisible(false)
{
}

NodeVisible::~NodeVisible(void)
{
}

bool NodeVisible::init()
{
	return true;
}

bool NodeVisible::detect()
{
	CCNode *pNode = SceneReader::sharedSceneReader()->getNodeByTag(_nTag);
	if (pNode != NULL && pNode->isVisible() == _bVisible)
	{
		return true;
	}
	return false;
}

void NodeVisible::serialize(const rapidjson::Value &val)
{
	int count = DICTOOL->getArrayCount_json(val, "dataitems");
	for (int i = 0; i < count; ++i)
	{
		const rapidjson::Value &subDict = DICTOOL->getSubDictionary_json(val, "dataitems", i);
		std::string key = DICTOOL->getStringValue_json(subDict, "key");
		if (key == "Tag")
		{
			_nTag = DICTOOL->getIntValue_json(subDict, "value");
			continue;
		}
		else if (key == "Visible")
		{
			_bVisible = DICTOOL->getIntValue_json(subDict, "value") != 0? true:false;
			continue;
		}
	}
}

void NodeVisible::serialize(cocos2d::extension::CocoLoader *pCocoLoader, cocos2d::extension::stExpCocoNode *pCocoNode)
{
	int length = pCocoNode->GetChildNum();
	int count = 0;
	stExpCocoNode *pDataItemsArray = pCocoNode->GetChildArray(pCocoLoader);
	std::string key;
	const char *str = NULL;
	for (int i = 0; i < length; ++i)
	{
		count = pDataItemsArray[i].GetChildNum();
		stExpCocoNode *pDataItemArray = pDataItemsArray[i].GetChildArray(pCocoLoader);
		key = pDataItemArray[0].GetValue(pCocoLoader);
		str = pDataItemArray[1].GetValue(pCocoLoader);
		if (key == "Tag")
		{
			if (str != NULL)
			{
				_nTag = atoi(str);//DICTOOL->getIntValue_json(subDict, "value");
			}
		}
		else if (key == "Visible")
		{
			if (str != NULL)
			{
				_bVisible = atoi(str) != 0? true:false;//DICTOOL->getIntValue_json(subDict, "value") != 0? true:false;
			}
		}
	}
}

void NodeVisible::removeAll()
{
	CCLOG("NodeVisible::removeAll");
}

File: tests/js-tests/src/CocoStudioTest/SceneTest/TriggerCode/Cons.js
/****************************************************************************
 Copyright (c) 2008-2010 Ricardo Quesada
 Copyright (c) 2011-2012 cocos2d-x.org
 Copyright (c) 2013-2014 Chukong Technologies Inc.

 http://www.cocos2d-x.org

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
 ****************************************************************************/
var TimeElapsed = ccs.BaseTriggerCondition.extend({
    _totalTime: 0,
    _tmpTime: 0,
    _scheduler: null,
    _success: false,
    ctor: function () {
        this._totalTime = 0;
        this._tmpTime = 0;
        this._scheduler = null;
        this._success = false;
        this._scheduler = cc.director.getScheduler();
    },
    
    init: function () {
        this._scheduler.scheduleCallbackForTarget(this, this.update);
        return true;
    },

    detect: function () {
        return this._success;
    },

    serialize: function (jsonVal) {
        var dataitems = jsonVal["dataitems"] || [];
        for (var i = 0; i < dataitems.length; i++) {
            var subDict = dataitems[i];
            var key = subDict["key"];
            if (key == "TotalTime") {
                this._totalTime = subDict["value"];
            }
        }
    },

    removeAll: function () {
        this._scheduler.unscheduleUpdateForTarget(this);
    },

    update: function (dt) {
        this._tmpTime += dt;
        if (this._tmpTime > this._totalTime) {
            this._tmpTime = 0.0;
            this._success = true;
        }
    }
});
var ArmatureActionState = ccs.BaseTriggerCondition.extend({
    _tag: -1,
    _state: -1,
    _success: false,
    _aniName: "",
    _comName: "",
    _armature: null,
    ctor: function () {
        this._tag = -1;
        this._state = -1;
        this._success = false;
        this._aniName = "";
        this._comName = "";
        this._armature = null;
    },
    
    init: function () {
        var node = ccs.sceneReader.getNodeByTag(this._tag);
        if (!node) return false;
        var render = node.getComponent(this._comName);
        if (!render) return false;
        var armature = render.getNode();
        if (!armature) return false;
        this._armature = armature;
        ccs.triggerManager.addArmatureMovementCallBack(this._armature, this.animationEvent, this);
        return true;
    },

    detect: function () {
        return this._success;
    },

    serialize: function (jsonVal) {
        var dataitems = jsonVal["dataitems"] || [];
        for (var i = 0; i < dataitems.length; i++) {
            var subDict = dataitems[i];
            var key = subDict["key"];
            if (key == "TotalTime") {
                this._totalTime = subDict["value"];
                continue;
            }
            if (key == "componentName") {
                this._comName = subDict["value"];
                continue;
            }
            if (key == "AnimationName") {
                this._aniName = subDict["value"];
                continue;
            }
            if (key == "ActionType") {
                this._state = subDict["value"];
            }
        }
    },

    removeAll: function () {
        if (this._armature) {
            ccs.triggerManager.removeArmatureMovementCallBack(this._armature, this.animationEvent, this);
        }
    },

    animationEvent: function (armature, movementType, movementID) {
        if (movementType == this._state && movementID == this._aniName) {
            this._success = true;
        }
    }
});
var NodeInRect = ccs.BaseTriggerCondition.extend({
    _tag: -1,
    _origin: null,
    _size: null,
    ctor: function () {
        this._tag = -1;
        this._origin = null;
        this._size = null;
        this._origin = cc.p(0, 0);
        this._size = cc.p(0, 0);
    },

    init: function () {

        return true;
    },

    detect: function () {
        var node = ccs.sceneReader.getNodeByTag(this._tag);
        if (node && Math.abs(node.x - this._origin.x) <= this._size.width && Math.abs(node.y - this._origin.y) <= this._size.height) {
            return true;
        }
        return false;
    },
    serialize: function (jsonVal) {
        var dataitems = jsonVal["dataitems"] || [];
        for (var i = 0; i < dataitems.length; i++) {
            var subDict = dataitems[i];
            var key = subDict["key"];
            if (key == "Tag") {
                this._tag = subDict["value"];
                continue;
            }
            if (key == "originX") {
                this._origin.x = subDict["value"];
                continue;
            }
            if (key == "originY") {
                this._origin.y = subDict["value"];
                continue;
            }
            if (key == "originX") {
                this._origin.x = subDict["value"];
                continue;
            }
            if (key == "originY") {
                this._origin.y = subDict["value"];
                continue;
            }
            if (key == "sizeWidth") {
                this._size.width = subDict["value"];
                continue;
            }
            if (key == "sizeHeight") {
                this._size.height = subDict["value"];
            }
        }
    },

    removeAll: function () {

    }
});

var NodeVisible = ccs.BaseTriggerCondition.extend({
    _tag: -1,
    _visible: false,
    cotr: function () {
        this._tag = -1;
        this._visible = false;
    },
    
    init: function () {
        return true;
    },

    detect: function () {
        var node = ccs.sceneReader.getNodeByTag(this._tag);
        if (node && node.visible == this._visible) {
            return true;
        }
        return false;
    },
    serialize: function (jsonVal) {
        var dataitems = jsonVal["dataitems"] || [];
        for (var i = 0; i < dataitems.length; i++) {
            var subDict = dataitems[i];
            var key = subDict["key"];
            if (key == "Tag") {
                this._tag = subDict["value"];
                continue;
            }
            if (key == "Visible") {
                this._visible = subDict["value"];
            }
        }
    },

    removeAll: function () {

    }
});


ccs.registerTriggerClass("TimeElapsed", TimeElapsed);
ccs.registerTriggerClass("ArmatureActionState", ArmatureActionState);
ccs.registerTriggerClass("NodeInRect", NodeInRect);
ccs.registerTriggerClass("NodeVisible", NodeVisible);